modules/twinklexfd.js

```diff
@@ -1616,6 +1703,28 @@ Twinkle.xfd.callback.evaluate = function(e) {
 			// find target and pass main as the callback
 			Twinkle.xfd.callbacks.rfd.findTarget(params, Twinkle.xfd.callbacks.rfd.main);
 			break;
+
+		case 'rm':
+			params = { reason: reason, newname: newname, rmtr: rmtr };
+			var talkpagename = mw.Title.newFromText(Morebits.pageNameNorm).getTalkPage().getPrefixedText();
+			Morebits.wiki.actionCompleted.redirect = rmtr ? 'Wikipedia:Requested moves/Technical requests' : talkpagename;
+			Morebits.wiki.actionCompleted.notice = 'Nomination completed, now redirecting to the discussion page';
+
+			if (rmtr) {
+				if (!newname.trim()) {
+					alert('Please enter the destination page name');
```

Comment:
Nice catch. Fixed it by adding an event listener to the rmtr checkbox that sets newname field as required or not depending on whether rmtr checkbox is checked or not. Browser-based required field validation is supported on all target browsers. 

This should actually be baked into Morebits.quickForm. It would eliminate a lot of form validation code from friendlytag ([L1996 ](https://github.com/azatoth/twinkle/blob/master/modules/friendlytag.js#L1996)and onwards), by making the browser do it for us. (Though this wouldn't have helped in this case, as required-ness here is conditional.)

Reply:
wtf is this magic?!  Glorious.  Not Twinkley, but glorious.  FWIW https://www.w3schools.com/jsref/prop_text_required.asp says it's not supported by Safari, but it works for me (v11.1.2).

I'd be curious what it would look like if indeed we validated everything (or as much as possible)with this rather than `alert`.  Can we customize the message or icon?  I'd be curious what @MusikAnimal thinks before you/we dive headfirst into it, but I think it's excellent here.

Reply:
> FWIW https://www.w3schools.com/jsref/prop_text_required.asp says it's not supported by Safari, but it works for me (v11.1.2).

I would not trust w3schools. They used to be [much, much worse](https://www.w3fools.com/), and could still contain inaccuracies. [MDN](https://developer.mozilla.org/en-US/) is probably the most well-respected resource in the industry, largely thanks to being a wiki. For browser compatibility, specifically, [caniuse.com](https://caniuse.com/) is generally accurate too and very comprehensive. In this case historical support for the `required` attribute isn't well-documented, but it's definitely been around for a long time. The [`:required` CSS pseudo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/:required) is said to have Safari support starting with version 5, so that pretty much answers your question. It should be safe to use.

Reply:
@Amorymeltzer Welcome to HTML5! The message [can be customised](https://dzone.com/articles/custom-validation-messages). Appearance/icon vary by browser and can't be customised, though I'm not sure if _that_ is supported everywhere. This is quite mainstream - even used in MediaWiki's [Special:UserLogin](https://en.wikipedia.org/w/index.php?title=Special:UserLogin) with the default message.

> it's not supported by Safari

Duh, yeah its [supported only on Safari 10.1+ if SO can be relied upon](https://stackoverflow.com/questions/23261301/required-attribute-not-work-in-safari-browser). I did recollect incorrectly that it's supported on all target browsers. I did this in #488, and indeed [a fallback was provided for Safari](https://github.com/siddharthvp/twinkle/blob/talktags/modules/friendlytag.js#L1727).

That being said, Safari is the only bad boy in the room, its supported everywhere else and this single field here isn't that important to validate, so we should be fine. 

If/when we do this in friendlytag, it may probably we worth writing a fallback.

Reply:
One of the StackOverflow answers references caniuse.com saying 10.1 was when it was added, but now as far as I can tell caniuse says the only known support is in version 13. But obviously older versions are supported, because it worked for Amory in 11.1. Regardless, both MDN and caniuse report the `:required` CSS pseudo-class is in Safari starting with version 5, and I don't know why this would hold true if the `required` attribute itself isn't supported. :man_shrugging:

I also have high doubts MediaWiki devs are actually testing things in Safari 5.1 (as claimed at [mw:Compatiblity#Browsers](https://www.mediawiki.org/wiki/Compatibility#Browsers)). That version is up to 8 years old, and even Apple doesn't support it anymore. I would wager we are OK to use `required`, especially given the [low usage of Safari](https://analytics.wikimedia.org/dashboards/browsers/#desktop-site-by-browser), but because editing is involved (as opposed to a read-only tool like XTools), maybe we should play it safe.

Reply:
> MDN and caniuse report the :required CSS pseudo-class is in Safari starting with version 5, and I don't know why this would hold true if the required attribute itself isn't supported. 

Maybe because it's easier to support the pseudo-class in CSS (simply defined as everything having the attribute) than actually supporting the functionality of it. 

The ability to actually check the input and report error was [added in Safari Technology Preview 19](https://webkit.org/blog/7099/html-interactive-form-validation/), which likely was first [applied in Safari v10.1](https://webkit.org/blog/7477/new-web-features-in-safari-10-1/) (it has the section "Interactive form validation"). 

> but because editing is involved (as opposed to a read-only tool like XTools), maybe we should play it safe.

I think we are safe enough. The problem arises only when someone ticks "Uncontroversial technical request" and submits without providing a page name, in this case {{RMassist}} will show a big red error, which will be displayed to the user - as WP:RM/TR is loaded a few seconds after action complete. They can thus fix the issue by editing the template. 

Complicating things here with something similar to `Twinkle.speedy.resolveCsdValues` is not warranted IMO. Anyway this level of validation is not commonplace on Twinkle - [the corresponding fix for merge tag target](https://github.com/azatoth/twinkle/commit/95e62e94358b51bd769b3942c0e4b790bb94b22e) is a very recent addition.

Reply:
Jah, I couldn't find anything on MDN while heading out the door ([there it is!](https://developer.mozilla.org/en-US/docs/Web/CSS/:required)), but caniuse is a great resource!

If it's not supported, will Safari silently fail or throw an error?  I think we've all reached the conclusion/consensus that whether support was added in 5 or 10 is murky, but even if we say it only works in 11+, per the analytics page that leaves behind 11% of desktop Safari users; if it's safe in 10, that's down to 6%.  With just under 6% of users using Safari, that's either 0.65% or 0.35% of all devices.  I'm inclined to think @MusikAnimal is right that we'll be fine, especially if it just silently skips.